packages/lesswrong/lib/types/collectionTypes.ts

```diff
@@ -254,4 +254,46 @@ interface SpotlightFirstPost {
   url: string;
 }
 
+// Sorry for declaring these so far from their function definitions. The
+// functions are defined in /server, and import cycles, etc.
+
+type CreateMutatorParams<T extends DbObject> = {
+  collection: CollectionBase<T>,
+  document: Partial<DbInsertion<T>>,
+  currentUser?: DbUser|null,
+  validate?: boolean,
+  context?: ResolverContext,
+};
+type CreateMutator = <T extends DbObject>(args: CreateMutatorParams<T>) => Promise<{data: T}>;
+
+type UpdateMutatorParams<T extends DbObject> = {
+  collection: CollectionBase<T>;
+  /** NB: Not technically required, you can replace with selector, though maybe
+   * you shouldn't */
+  documentId: string;
+  /** TODO: Really selector should be allowed to replace documentId. Although
+   * this is also a caveat for deprecation */
+  selector?: MongoSelector<T>,
+  data?: Partial<DbInsertion<T>>;
+  set?: Partial<DbInsertion<T>>;
+  unset?: any;
+  currentUser?: DbUser | null;
+  validate?: boolean;
+  context?: ResolverContext;
+  document?: T | null;
+};
+type UpdateMutator = <T extends DbObject>(args: UpdateMutatorParams<T>) => Promise<{ data: T }>;
```

Comment:
Slightly verbose but will this do the trick to fix the "selector or id" issue?
```suggestion
type UpdateMutatorParamsBase<T extends DbObject> = {
  collection: CollectionBase<T>;
  data?: Partial<DbInsertion<T>>;
  set?: Partial<DbInsertion<T>>;
  unset?: any;
  currentUser?: DbUser | null;
  validate?: boolean;
  context?: ResolverContext;
  document?: T | null;
};
type UpdateMutatorParamsWithDocId<T extends DbObject> = UpdateMutatorParamsBase<T> & {documentId: string, selector?: never};
type UpdateMutatorParamsWithSelector<T extends DbObject> = UpdateMutatorParamsBase<T> & {documentId?: never, selector: MongoSelector<T>};
type UpdateMutatorParams<T extends DbObject> = UpdateMutatorParamsWithDocId<T> | UpdateMutatorParamsWithSelector<T>;

type UpdateMutator = <T extends DbObject>(args: UpdateMutatorParams<T>) => Promise<{ data: T }>;
```
I got the `never` trick from [here](https://stackoverflow.com/questions/37688318/typescript-interface-possible-to-make-one-or-the-other-properties-required), I think you could also use undefined

Reply:
This does catch one place (introduced by me :( ) where both are used when only one is needed

Reply:
Yeah, maybe this is good. Maybe we should keep selector around tbh.